File: in_Java/0074 Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) return false;
        int row = matrix.length, col = matrix[0].length;
        int id = 0;
        for(int i = 0; i < row; i++){
            if(matrix[i][col-1] == target) return true;
            if(matrix[i][col-1] > target){
                id = i;
                break;
            }
        }
        int[] arr = matrix[id];
        int left = 0, right = col-1;
        while(left <= right){
            int m = left + (right-left)/2;
            if(arr[m] == target) return true;
            else if(arr[m] < target) left = m+1;
            else right = m-1;
        }
        return false;
    }
}

File: in_Python/0074 Search a 2D Matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if matrix == None:
            return False
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        row = len(matrix)
        col = len(matrix[0])
        rid = 0
        for r in range(row):
            if matrix[r][col-1] >= target:
                rid = r
                break
        r = matrix[rid]
        i = 0
        j = col - 1
        while i <= j:
            m = i + (j-i) / 2
            if r[m] == target:
                return True
            elif r[m] > target:
                j = m - 1
            else:
                i = m + 1
        return False
        